File: frontend/src/api/metrics.ts
import { api } from './client'
import type {
  AppVersionAdoptionResponse,
  AppVersionSeriesResponse,
  BreakdownTimeline,
  EventMetricBreakdownsResponse,
  EventMetricsResponse,
  DistributionDriftsResponse,
  EventWindowMetrics,
  MonitoringSignal,
  OverviewKpiSeries,
  ReleaseRegressionsResponse,
  SeasonalityHeatmap,
  TopEvent,
  TopMoverItem,
} from '../types'

export interface EventsMetricsParams {
  event_type_id?: string
  search?: string
  status?: string[]
  tag?: string
  from?: string
  to?: string
}

export const metricsApi = {
  getEventsMetrics: (slug: string, params?: EventsMetricsParams) => {
    const sp = new URLSearchParams()
    if (params?.event_type_id) sp.set('event_type_id', params.event_type_id)
    if (params?.search) sp.set('search', params.search)
    if (params?.status?.length) {
      for (const s of params.status) sp.append('status', s)
    }
    if (params?.tag) sp.set('tag', params.tag)
    if (params?.from) sp.set('from', params.from)
    if (params?.to) sp.set('to', params.to)
    const qs = sp.toString()
    return api.get<EventMetricsResponse>(`/projects/${slug}/events-metrics${qs ? `?${qs}` : ''}`)
  },

  getProjectTotalMetrics: (
    slug: string,
    params?: { scan_config_id?: string; from?: string; to?: string },
  ) => {
    const sp = new URLSearchParams()
    if (params?.scan_config_id) sp.set('scan_config_id', params.scan_config_id)
    if (params?.from) sp.set('from', params.from)
    if (params?.to) sp.set('to', params.to)
    const qs = sp.toString()
    return api.get<EventMetricsResponse>(`/projects/${slug}/metrics/total${qs ? `?${qs}` : ''}`)
  },

  getEventMetrics: (slug: string, eventId: string, params?: { from?: string; to?: string }) => {
    const sp = new URLSearchParams()
    if (params?.from) sp.set('from', params.from)
    if (params?.to) sp.set('to', params.to)
    const qs = sp.toString()
    return api.get<EventMetricsResponse>(`/projects/${slug}/events/${eventId}/metrics${qs ? `?${qs}` : ''}`)
  },

  getEventMetricBreakdowns: (
    slug: string,
    eventId: string,
    params?: { column?: string; from?: string; to?: string },
  ) => {
    const sp = new URLSearchParams()
    if (params?.column) sp.set('column', params.column)
    if (params?.from) sp.set('from', params.from)
    if (params?.to) sp.set('to', params.to)
    const qs = sp.toString()
    return api.get<EventMetricBreakdownsResponse>(
      `/projects/${slug}/events/${eventId}/metrics/breakdowns${qs ? `?${qs}` : ''}`,
    )
  },

  getEventTypeMetrics: (slug: string, eventTypeId: string, params?: { from?: string; to?: string }) => {
    const sp = new URLSearchParams()
    if (params?.from) sp.set('from', params.from)
    if (params?.to) sp.set('to', params.to)
    const qs = sp.toString()
    return api.get<EventMetricsResponse>(`/projects/${slug}/event-types/${eventTypeId}/metrics${qs ? `?${qs}` : ''}`)
  },

  getEventsWindowMetrics: (
    slug: string,
    data: { event_ids: string[]; time_from: string; time_to: string },
  ) => api.post<EventWindowMetrics[]>(`/projects/${slug}/events/window-metrics`, data),

  getActiveSignals: (slug: string, eventIds?: string[], opts?: { expanded?: boolean }) => {
    // Empty / no ids → cacheable GET. Any filter → POST body, because with
    // 500+ events we'd otherwise blow past proxy/browser query-string limits.
    if (!eventIds || eventIds.length === 0) {
      // expanded=true (AnomaliesPage) adds per-event scopes and keeps tagged
      // incident children; the other callers omit it for the collapsed rollup.
      const query = opts?.expanded ? '?expanded=true' : ''
      return api.get<MonitoringSignal[]>(`/projects/${slug}/anomalies/signals${query}`)
    }
    return api.post<MonitoringSignal[]>(
      `/projects/${slug}/anomalies/signals/query`,
      { event_ids: eventIds },
    )
  },

  getTopMovers: (
    slug: string,
    scanConfigId: string,
    params: { scope_type: string; scope_ref: string; bucket: string; limit?: number },
  ) => {
    const sp = new URLSearchParams({
      scope_type: params.scope_type,
      scope_ref: params.scope_ref,
      bucket: params.bucket,
    })
    if (params.limit !== undefined) sp.set('limit', String(params.limit))
    return api.get<TopMoverItem[]>(
      `/projects/${slug}/scans/${scanConfigId}/top-movers?${sp.toString()}`,
    )
  },

  getOverviewKpiSeries: (slug: string, days = 14) =>
    api.get<OverviewKpiSeries>(`/projects/${slug}/overview/kpi-series?days=${days}`),

  getTopEvents: (slug: string, params?: { windowHours?: number; limit?: number }) => {
    const sp = new URLSearchParams()
    if (params?.windowHours !== undefined) sp.set('window_hours', String(params.windowHours))
    if (params?.limit !== undefined) sp.set('limit', String(params.limit))
    const qs = sp.toString()
    return api.get<TopEvent[]>(`/projects/${slug}/overview/top-events${qs ? `?${qs}` : ''}`)
  },

  getSeasonalityHeatmap: (
    slug: string,
    scanConfigId: string,
    params: { scope_type: string; scope_ref: string; from?: string; to?: string },
  ) => {
    const sp = new URLSearchParams({
      scope_type: params.scope_type,
      scope_ref: params.scope_ref,
    })
    if (params.from) sp.set('from', params.from)
    if (params.to) sp.set('to', params.to)
    return api.get<SeasonalityHeatmap>(
      `/projects/${slug}/scans/${scanConfigId}/seasonality?${sp.toString()}`,
    )
  },

  getBreakdownTimeline: (
    slug: string,
    scanConfigId: string,
    params: {
      scope_type: string
      scope_ref: string
      breakdown_column: string
      breakdown_value: string
      is_other?: boolean
      from?: string
      to?: string
    },
  ) => {
    const sp = new URLSearchParams({
      scope_type: params.scope_type,
      scope_ref: params.scope_ref,
      breakdown_column: params.breakdown_column,
      breakdown_value: params.breakdown_value,
    })
    if (params.is_other !== undefined) sp.set('is_other', String(params.is_other))
    if (params.from) sp.set('from', params.from)
    if (params.to) sp.set('to', params.to)
    return api.get<BreakdownTimeline>(
      `/projects/${slug}/scans/${scanConfigId}/breakdown-timeline?${sp.toString()}`,
    )
  },

  getDistributionDrifts: (
    slug: string,
    params: {
      scope_type: 'project_total' | 'event_type' | 'event'
      scope_ref: string
      scan_config_id?: string
      from?: string
      to?: string
    },
  ) => {
    const sp = new URLSearchParams({
      scope_type: params.scope_type,
      scope_ref: params.scope_ref,
    })
    if (params.scan_config_id) sp.set('scan_config_id', params.scan_config_id)
    if (params.from) sp.set('from', params.from)
    if (params.to) sp.set('to', params.to)
    return api.get<DistributionDriftsResponse>(
      `/projects/${slug}/distribution-drifts?${sp.toString()}`,
    )
  },

  getAppVersionSeries: (
    slug: string,
    scanConfigId: string,
    params?: {
      scope_type?: 'project_total' | 'event_type' | 'event'
      scope_ref?: string
      from?: string
      to?: string
    },
  ) => {
    const sp = new URLSearchParams()
    if (params?.scope_type) sp.set('scope_type', params.scope_type)
    if (params?.scope_ref) sp.set('scope_ref', params.scope_ref)
    if (params?.from) sp.set('from', params.from)
    if (params?.to) sp.set('to', params.to)
    const qs = sp.toString()
    return api.get<AppVersionSeriesResponse>(
      `/projects/${slug}/scans/${scanConfigId}/app-versions${qs ? `?${qs}` : ''}`,
    )
  },

  getAppVersionAdoption: (
    slug: string,
    scanConfigId: string,
    params?: { from?: string; to?: string },
  ) => {
    const sp = new URLSearchParams()
    if (params?.from) sp.set('from', params.from)
    if (params?.to) sp.set('to', params.to)
    const qs = sp.toString()
    return api.get<AppVersionAdoptionResponse>(
      `/projects/${slug}/scans/${scanConfigId}/version-adoption${qs ? `?${qs}` : ''}`,
    )
  },

  getReleaseRegressions: (
    slug: string,
    scanConfigId: string,
    params?: { scope_type?: 'event' | 'event_type' },
  ) => {
    const sp = new URLSearchParams()
    if (params?.scope_type) sp.set('scope_type', params.scope_type)
    const qs = sp.toString()
    return api.get<ReleaseRegressionsResponse>(
      `/projects/${slug}/scans/${scanConfigId}/release-regressions${qs ? `?${qs}` : ''}`,
    )
  },
}

File: backend/src/tripl/api/v1/metrics.py
import uuid
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Query

from tripl.api.deps import SessionDep
from tripl.models.domain_enums import MetricScopeType
from tripl.models.event import EventStatus
from tripl.schemas.event_metric import (
    ActiveSignalsQuery,
    AppVersionAdoptionResponse,
    AppVersionSeriesResponse,
    BreakdownTimelineResponse,
    DistributionDriftsResponse,
    EventMetricBreakdownsResponse,
    EventMetricsResponse,
    EventWindowMetricsRequest,
    EventWindowMetricsResponse,
    MetricSignalResponse,
    OverviewKpiSeriesResponse,
    ReleaseRegressionsResponse,
    SeasonalityHeatmapResponse,
    TopEventResponse,
    TopMoverItem,
)
from tripl.services import metrics_insights_service, metrics_service

router = APIRouter(tags=["metrics"])

TimeFrom = Annotated[datetime | None, Query(alias="from")]
TimeTo = Annotated[datetime | None, Query(alias="to")]
EventIds = Annotated[list[uuid.UUID] | None, Query(alias="event_id")]
# Every ``scope_type`` below is declared MetricScopeType rather than str: the
# value is bound against a native Postgres ``metric_scope_type`` column, so an
# out-of-enum string used to reach the driver and surface as a 500 instead of a
# 422 (tripl-57g0). Spelling out the full enum is deliberate — the endpoints
# that support only a subset (app-versions, distribution-drifts) already raise
# their own 4xx for the scopes they don't handle, so this rejects garbage at the
# edge without narrowing anything that works today.


@router.get(
    "/projects/{slug}/events-metrics",
    response_model=EventMetricsResponse,
)
async def get_events_metrics(
    session: SessionDep,
    slug: str,
    event_type_id: uuid.UUID | None = None,
    search: str | None = None,
    tag: str | None = None,
    # list[EventStatus] (not list[str]): the filter lands in Event.status.in_(),
    # a native Postgres enum, so one bad member 500'd the whole request. Mirrors
    # GET /events?status= — same filter, same enum, now the same 422 (tripl-57g0).
    status: Annotated[list[EventStatus] | None, Query()] = None,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> EventMetricsResponse:
    return await metrics_service.get_events_metrics(
        session,
        slug,
        event_type_id=event_type_id,
        search=search,
        tag=tag,
        status=[member.value for member in status] if status else None,
        time_from=time_from,
        time_to=time_to,
    )


@router.post(
    "/projects/{slug}/events/window-metrics",
    response_model=list[EventWindowMetricsResponse],
)
async def get_events_window_metrics(
    session: SessionDep,
    slug: str,
    data: EventWindowMetricsRequest,
) -> list[EventWindowMetricsResponse]:
    return await metrics_service.get_events_window_metrics(
        session,
        slug,
        event_ids=data.event_ids,
        time_from=data.time_from,
        time_to=data.time_to,
    )


@router.get(
    "/projects/{slug}/overview/kpi-series",
    response_model=OverviewKpiSeriesResponse,
)
async def get_overview_kpi_series(
    session: SessionDep,
    slug: str,
    days: int = Query(14, ge=1, le=365),
) -> OverviewKpiSeriesResponse:
    return await metrics_service.get_overview_kpi_series(session, slug, days=days)


@router.get(
    "/projects/{slug}/overview/top-events",
    response_model=list[TopEventResponse],
)
async def get_overview_top_events(
    session: SessionDep,
    slug: str,
    window_hours: int = Query(48, ge=1, le=720),
    limit: int = Query(6, ge=1, le=100),
) -> list[TopEventResponse]:
    return await metrics_service.get_top_events_by_volume(
        session,
        slug,
        window_hours=window_hours,
        limit=limit,
    )


@router.get(
    "/projects/{slug}/metrics/total",
    response_model=EventMetricsResponse,
)
async def get_project_total_metrics(
    session: SessionDep,
    slug: str,
    scan_config_id: uuid.UUID | None = None,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> EventMetricsResponse:
    return await metrics_service.get_project_total_metrics(
        session,
        slug,
        scan_config_id=scan_config_id,
        time_from=time_from,
        time_to=time_to,
    )


@router.get(
    "/projects/{slug}/events/{event_id}/metrics",
    response_model=EventMetricsResponse,
)
async def get_event_metrics(
    session: SessionDep,
    slug: str,
    event_id: uuid.UUID,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> EventMetricsResponse:
    return await metrics_service.get_event_metrics(session, slug, event_id, time_from, time_to)


@router.get(
    "/projects/{slug}/events/{event_id}/metrics/breakdowns",
    response_model=EventMetricBreakdownsResponse,
)
async def get_event_metric_breakdowns(
    session: SessionDep,
    slug: str,
    event_id: uuid.UUID,
    column: str | None = None,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> EventMetricBreakdownsResponse:
    return await metrics_service.get_event_metric_breakdowns(
        session,
        slug,
        event_id,
        column=column,
        time_from=time_from,
        time_to=time_to,
    )


@router.get(
    "/projects/{slug}/event-types/{event_type_id}/metrics",
    response_model=EventMetricsResponse,
)
async def get_event_type_metrics(
    session: SessionDep,
    slug: str,
    event_type_id: uuid.UUID,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> EventMetricsResponse:
    return await metrics_service.get_event_type_metrics(
        session, slug, event_type_id, time_from, time_to
    )


@router.get(
    "/projects/{slug}/anomalies/signals",
    response_model=list[MetricSignalResponse],
)
async def get_active_signals(
    session: SessionDep,
    slug: str,
    event_ids: EventIds = None,
    expanded: bool = False,
) -> list[MetricSignalResponse]:
    """Cacheable no-args variant. For filtering by a large event-id list
    (>>a few), prefer ``POST /anomalies/signals/query`` — GET's query-string
    overflow is real once you cross ~50 ids (proxy/browser limits).

    ``expanded=true`` (the AnomaliesPage view) also surfaces per-event scopes
    and keeps each incident's child rows, tagged ``incident_child`` rather than
    collapsed into the parent project_total signal."""
    return await metrics_insights_service.get_active_signals(
        session, slug, event_ids=event_ids, expanded=expanded
    )


@router.post(
    "/projects/{slug}/anomalies/signals/query",
    response_model=list[MetricSignalResponse],
)
async def query_active_signals(
    session: SessionDep,
    slug: str,
    data: ActiveSignalsQuery,
) -> list[MetricSignalResponse]:
    return await metrics_insights_service.get_active_signals(
        session, slug, event_ids=data.event_ids or None
    )


@router.get(
    "/projects/{slug}/scans/{scan_config_id}/top-movers",
    response_model=list[TopMoverItem],
)
async def get_top_movers(
    session: SessionDep,
    slug: str,
    scan_config_id: uuid.UUID,
    scope_type: MetricScopeType,
    scope_ref: str,
    bucket: datetime,
    limit: int = Query(10, ge=1, le=100),
) -> list[TopMoverItem]:
    """Top-N breakdown rows that "moved" a given anomaly bucket, |z| desc."""
    return await metrics_insights_service.get_top_movers(
        session,
        slug,
        scan_config_id=scan_config_id,
        scope_type=scope_type,
        scope_ref=scope_ref,
        bucket=bucket,
        limit=limit,
    )


@router.get(
    "/projects/{slug}/scans/{scan_config_id}/seasonality",
    response_model=SeasonalityHeatmapResponse,
)
async def get_seasonality_heatmap(
    session: SessionDep,
    slug: str,
    scan_config_id: uuid.UUID,
    scope_type: MetricScopeType,
    scope_ref: str,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> SeasonalityHeatmapResponse:
    """7×24 hour-of-day × weekday heatmap of volume and anomaly density."""
    return await metrics_insights_service.get_seasonality_heatmap(
        session,
        slug,
        scan_config_id=scan_config_id,
        scope_type=scope_type,
        scope_ref=scope_ref,
        time_from=time_from,
        time_to=time_to,
    )


@router.get(
    "/projects/{slug}/scans/{scan_config_id}/breakdown-timeline",
    response_model=BreakdownTimelineResponse,
)
async def get_breakdown_timeline(
    session: SessionDep,
    slug: str,
    scan_config_id: uuid.UUID,
    scope_type: MetricScopeType,
    scope_ref: str,
    breakdown_column: str,
    breakdown_value: str,
    is_other: bool = False,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> BreakdownTimelineResponse:
    """Per-bucket count timeline for one breakdown_value (drill-down)."""
    return await metrics_insights_service.get_breakdown_timeline(
        session,
        slug,
        scan_config_id=scan_config_id,
        scope_type=scope_type,
        scope_ref=scope_ref,
        breakdown_column=breakdown_column,
        breakdown_value=breakdown_value,
        is_other=is_other,
        time_from=time_from,
        time_to=time_to,
    )


@router.get(
    "/projects/{slug}/scans/{scan_config_id}/app-versions",
    response_model=AppVersionSeriesResponse,
)
async def get_app_version_series(
    session: SessionDep,
    slug: str,
    scan_config_id: uuid.UUID,
    scope_type: MetricScopeType = MetricScopeType.project_total,
    scope_ref: str | None = None,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> AppVersionSeriesResponse:
    return await metrics_service.get_app_version_series(
        session,
        slug,
        scan_config_id=scan_config_id,
        scope_type=scope_type,
        scope_ref=scope_ref,
        time_from=time_from,
        time_to=time_to,
    )


@router.get(
    "/projects/{slug}/scans/{scan_config_id}/version-adoption",
    response_model=AppVersionAdoptionResponse,
)
async def get_app_version_adoption(
    session: SessionDep,
    slug: str,
    scan_config_id: uuid.UUID,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> AppVersionAdoptionResponse:
    return await metrics_service.get_app_version_adoption(
        session,
        slug,
        scan_config_id=scan_config_id,
        time_from=time_from,
        time_to=time_to,
    )


@router.get(
    "/projects/{slug}/scans/{scan_config_id}/release-regressions",
    response_model=ReleaseRegressionsResponse,
)
async def get_release_regressions(
    session: SessionDep,
    slug: str,
    scan_config_id: uuid.UUID,
    scope_type: MetricScopeType | None = None,
) -> ReleaseRegressionsResponse:
    return await metrics_service.get_release_regressions(
        session,
        slug,
        scan_config_id=scan_config_id,
        scope_type=scope_type,
    )


@router.get(
    "/projects/{slug}/distribution-drifts",
    response_model=DistributionDriftsResponse,
)
async def get_distribution_drifts(
    session: SessionDep,
    slug: str,
    scope_type: MetricScopeType,
    scope_ref: str,
    scan_config_id: uuid.UUID | None = None,
    time_from: TimeFrom = None,
    time_to: TimeTo = None,
) -> DistributionDriftsResponse:
    return await metrics_insights_service.get_distribution_drifts(
        session,
        slug,
        scope_type=scope_type,
        scope_ref=scope_ref,
        scan_config_id=scan_config_id,
        time_from=time_from,
        time_to=time_to,
    )
